[CI] RetentionLeaseIT testCanRenewRetentionLeaseWithoutWaitingForShards failed

`RetentionLeaseIT testCanRenewRetentionLeaseWithoutWaitingForShards` failed in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+g1gc/565/console

The exception that caused the failure was:

```
java.lang.AssertionError: ElasticsearchException[java.io.IOException: failed to read /var/lib/jenkins/workspace/elastic+elasticsearch+master+g1gc/server/build/testrun/integTest/temp/org.elasticsearch.index.seqno.RetentionLeaseIT_C70B235D5835B1AE-001/tempDir-008/data/nodes/0/indices/3YlDl4joQjiLzNdRZJyHpg/0/_state/retention-leases-1.st]; nested: IOException[failed to read /var/lib/jenkins/workspace/elastic+elasticsearch+master+g1gc/server/build/testrun/integTest/temp/org.elasticsearch.index.seqno.RetentionLeaseIT_C70B235D5835B1AE-001/tempDir-008/data/nodes/0/indices/3YlDl4joQjiLzNdRZJyHpg/0/_state/retention-leases-1.st]; nested: NoSuchFileException[/var/lib/jenkins/workspace/elastic+elasticsearch+master+g1gc/server/build/testrun/integTest/temp/org.elasticsearch.index.seqno.RetentionLeaseIT_C70B235D5835B1AE-001/tempDir-008/data/nodes/0/indices/3YlDl4joQjiLzNdRZJyHpg/0/_state/retention-leases-1.st];
```

The full stack traces of those underlying failures are not available as the assertion that made the test fail was:

```
                    } catch (final Exception e) {
                        fail(e.toString());
                    }
```

The problem does not reproduce locally using:

```
./gradlew :server:integTest --tests "org.elasticsearch.index.seqno.RetentionLeaseIT.testCanRenewRetentionLeaseWithoutWaitingForShards" -Dtests.seed=C70B235D5835B1AE -Dtests.security.manager=true -Dtests.jvm.argline="-XX:-UseConcMarkSweepGC -XX:+UseG1GC" -Dtests.locale=ms-BN -Dtests.timezone=Europe/Vilnius -Dcompiler.java=12 -Druntime.java=11
```

The stack traces added in #42425 clarify the situation:

```
org.elasticsearch.index.seqno.RetentionLeaseIT > testCanRenewRetentionLeaseWithoutWaitingForShards FAILED
    java.lang.AssertionError: unexpected
        at __randomizedtesting.SeedInfo.seed([741596F11316B18F:6ED407F2498426AD]:0)
        at org.elasticsearch.index.seqno.RetentionLeaseIT.failWithException(RetentionLeaseIT.java:675)
        at org.elasticsearch.index.seqno.RetentionLeaseIT.lambda$testCanRenewRetentionLeaseWithoutWaitingForShards$25(RetentionLeaseIT.java:598)
        at org.elasticsearch.index.seqno.RetentionLeaseIT.runWaitForShardsTest(RetentionLeaseIT.java:671)
        at org.elasticsearch.index.seqno.RetentionLeaseIT.testCanRenewRetentionLeaseWithoutWaitingForShards(RetentionLeaseIT.java:580)

        Caused by:
        java.lang.IllegalStateException: unable to find state files with generation id 1 returned by findMaxGenerationId function, in data folders [/var/lib/jenkins/workspace/elastic+elasticsearch+master+multijob-unix-compatibility/os/ubuntu-16.04&&immutable/server/build/testrun/integTest/temp/org.elasticsearch.index.seqno.RetentionLeaseIT_741596F11316B18F-001/tempDir-013/node_t1/nodes/0/indices/BRVbjjL7SDiL6xe0VnQ_gw/0], concurrent writes?
            at org.elasticsearch.gateway.MetaDataStateFormat.loadLatestStateWithGeneration(MetaDataStateFormat.java:439)
            at org.elasticsearch.gateway.MetaDataStateFormat.loadLatestState(MetaDataStateFormat.java:454)
            at org.elasticsearch.index.seqno.ReplicationTracker.loadRetentionLeases(ReplicationTracker.java:349)
            at org.elasticsearch.index.shard.IndexShard.loadRetentionLeases(IndexShard.java:2015)
            at org.elasticsearch.index.seqno.RetentionLeaseIT.lambda$testCanRenewRetentionLeaseWithoutWaitingForShards$24(RetentionLeaseIT.java:596)
            at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:850)
            at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:836)
            at org.elasticsearch.index.seqno.RetentionLeaseIT.lambda$testCanRenewRetentionLeaseWithoutWaitingForShards$25(RetentionLeaseIT.java:596)
            ... 2 more
```

I opened #42620.

--- a/server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java
+++ b/server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java
@@ -346,7 +346,10 @@ public class ReplicationTracker extends AbstractIndexShardComponent implements L
      * @throws IOException if an I/O exception occurs reading the retention leases
      */
     public RetentionLeases loadRetentionLeases(final Path path) throws IOException {
-        final RetentionLeases retentionLeases = RetentionLeases.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY, path);
+        final RetentionLeases retentionLeases;
+        synchronized (retentionLeasePersistenceLock) {
+            retentionLeases = RetentionLeases.FORMAT.loadLatestState(logger, NamedXContentRegistry.EMPTY, path);
+        }
 
         // TODO after backporting we expect this never to happen in 8.x, so adjust this to throw an exception instead.
         assert Version.CURRENT.major <= 8 : "throw an exception instead of returning EMPTY on null";

